### PR TITLE
Update to use full equations of motion

### PR DESCRIPTION
**Pull request by brocksam**

### CODE REVIEW

File: src/generate_eom.py [hunk 1]
```diff
@@ -707,17 +747,8 @@ def gen_eom_for_opty(steer_with=SteerWith.MUSCLES, include_roll_torque=False):
...
-
-    contact_x = sm.Matrix([q1, q2, u1, u2])
-    contact_eom = sm.Matrix([
-        q1.diff(t) - u1, q2.diff(t) - u2,
-        u1.diff(t) - u1p_def, u2.diff(t) - u2p_def
-    ])
-
-    x = dyn_x.col_join(contact_x)
-    eom = dyn_eom.col_join(contact_eom)
+    x = kane.q.col_join(kane.u)
```
Comment (moorepants) [inline src/generate_eom.py]: I guess this ordering will not matter for opty, so if you want to make x have the q's and u's numerically ordered you could just to `q.col_join(u)`.

File: src/generate_eom.py [hunk 2]
```diff
@@ -707,17 +747,8 @@ def gen_eom_for_opty(steer_with=SteerWith.MUSCLES, include_roll_torque=False):
...
+    eom = kane.mass_matrix_full*x - kane.forcing_full
```
Comment (moorepants) [inline src/generate_eom.py]: It does matter in this equation though, so you have to use the x as defined above for this multiplication, but you don't need x in that order to pass to opty.

**Comment by moorepants** [comment]
Might be worth a try to increase the sympy cache `SYMPY_CACHE_SIZE=2000`.

**Comment by moorepants** [comment]
I was trying to run this but am getting:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File ~/Manuscripts/muscle-driven-bicycle-paper/src/test_jac.py:4
      1 from generate_eom import gen_eom_for_opty
      2 from utils import csejac
----> 4 data = gen_eom_for_opty()
      6 csejac(data.eom, data.x)
      8 data.eom.jacobian(data.x)

File ~/Manuscripts/muscle-driven-bicycle-paper/src/generate_eom.py:534, in gen_eom_for_opty(steer_with, include_roll_torque)
    531 tau_a, tau_d, b_tanh = sm.symbols('tau_a, tau_d, b_tanh')
    533 bicep_right_pathway = mec.LinearPathway(gm, hm)
--> 534 bicep_right_activation = FirstOrderActivationDeGroote2016.with_defaults(
    535     'bi_r',
    536     activation_time_constant=tau_a,
    537     deactivation_time_constant=tau_d,
    538     smoothing_rate=b_tanh,
    539 )
    540 bicep_right = MusculotendonDeGroote2016(
    541     'bi_r',
    542     bicep_right_pathway,
   (...)
    549     fiber_damping_coefficient=beta,
    550 )
    552 bicep_left_pathway = mec.LinearPathway(im, jm)

TypeError: FirstOrderActivationDeGroote2016.with_defaults() got an unexpected keyword argument 'activation_time_constant'
```

**Comment by brocksam** [comment]
> I was trying to run this but am getting:
> 
> 
> 
> ```
> 
> ---------------------------------------------------------------------------
> 
> TypeError                                 Traceback (most recent call last)
> 
> File ~/Manuscripts/muscle-driven-bicycle-paper/src/test_jac.py:4
> 
>       1 from generate_eom import gen_eom_for_opty
> 
>       2 from utils import csejac
> 
> ----> 4 data = gen_eom_for_opty()
> 
>       6 csejac(data.eom, data.x)
> 
>       8 data.eom.jacobian(data.x)
> 
> 
> 
> File ~/Manuscripts/muscle-driven-bicycle-paper/src/generate_eom.py:534, in gen_eom_for_opty(steer_with, include_roll_torque)
> 
>     531 tau_a, tau_d, b_tanh = sm.symbols('tau_a, tau_d, b_tanh')
> 
>     533 bicep_right_pathway = mec.LinearPathway(gm, hm)
> 
> --> 534 bicep_right_activation = FirstOrderActivationDeGroote2016.with_defaults(
> 
>     535     'bi_r',
> 
>     536     activation_time_constant=tau_a,
> 
>     537     deactivation_time_constant=tau_d,
> 
>     538     smoothing_rate=b_tanh,
> 
>     539 )
> 
>     540 bicep_right = MusculotendonDeGroote2016(
> 
>     541     'bi_r',
> 
>     542     bicep_right_pathway,
> 
>    (...)
> 
>     549     fiber_damping_coefficient=beta,
> 
>     550 )
> 
>     552 bicep_left_pathway = mec.LinearPathway(im, jm)
> 
> 
> 
> TypeError: FirstOrderActivationDeGroote2016.with_defaults() got an unexpected keyword argument 'activation_time_constant'
> 
> ```

Should just be using the normal `__init__`, not `with_constants` constructor.

**Comment by brocksam** [comment]
@moorepants I just pushed commit with the fix so you should now be able to generate the EoMs with the muscles.

**Comment by brocksam** [comment]
Using the branch from this opty PR https://github.com/csu-hmc/opty/pull/102 we can now get through the Jacobian derivation and compilation. It takes about 3 minutes on my laptop to get through everything and hit Ipopt. Sadly Ipopt immediate fails with the message `EXIT: Invalid number in NLP function or derivative detected.`

**Comment by moorepants** [comment]
Did you manually check `prob.con(initial_guess)` and `prob.jac(initial_guess)`?

**Comment by moorepants** [comment]
Also are you testing the jacobian's without the muscles? So only steer torque?

**Comment by brocksam** [comment]
> Did you manually check `prob.con(initial_guess)` and `prob.jac(initial_guess)`? 

No, but this sounds like a good suggestion of where to start investigating this morning, thanks. Especially as we're currently using an array of zeros for the initial guess, could easily be a floating point division by zero or similar somewhere. Will also look directly at the generated C code.

> Also are you testing the jacobian's without the muscles? So only steer torque?

I've tried both, both immediate kill Ipopt. So that likely means is something common is a term associated with the EoMs. I'm starting to investigate with steer torque only.

**Comment by moorepants** [comment]
The error seems to be in the objective:
```
EXIT: Invalid number in NLP function or derivative detected.
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
File ~/Manuscripts/muscle-driven-bicycle-paper/src/opty_ocp.py:152
    149 initial_guess = np.zeros(problem.num_free)
    151 # Find the optimal solution.
--> 152 sol, info = problem.solve(initial_guess)
    154 q1_sol = sol[:NUM_NODES]
    155 q2_sol = sol[NUM_NODES:2*NUM_NODES]

File ~/miniconda/envs/muscle-bike-paper/lib/python3.11/site-packages/cyipopt/cython/ipopt_wrapper.pyx:642, in ipopt_wrapper.Problem.solve()

File ~/miniconda/envs/muscle-bike-paper/lib/python3.11/site-packages/cyipopt/cython/ipopt_wrapper.pyx:700, in ipopt_wrapper.gradient_cb()

File ~/src/opty/opty/direct_collocation.py:226, in Problem.gradient(self, free)
    203 """Returns the value of the gradient of the objective function given a
    204 solution to the problem.
    205 
   (...)
    223 
    224 """
    225 # This should return a column vector.
--> 226 return self.obj_grad(free)

File ~/Manuscripts/muscle-driven-bicycle-paper/src/opty_ocp.py:39, in obj_grad(free)
     37 q2 = free[NUM_NODES:2*NUM_NODES]
     38 grad = np.zeros_like(free)
---> 39 dJdq1 = np.pi*LATERAL_DISPLACEMENT*(LATERAL_DISPLACEMENT*(1 - cos(np.pi*q1/LONGITUDINAL_DISPLACEMENT))/2 - q2)*sin(pi*q1/LONGITUDINAL_DISPLACEMENT)/LONGITUDINAL_DISPLACEMENT
     40 dJdq2 = LATERAL_DISPLACEMENT*(cos(pi*q1/LONGITUDINAL_DISPLACEMENT) - 1) + 2*q2
     41 grad[:NUM_NODES] = dJdq1

NameError: name 'cos' is not defined
> /home/moorepants/Manuscripts/muscle-driven-bicycle-paper/src/opty_ocp.py(39)obj_grad()
     37     q2 = free[NUM_NODES:2*NUM_NODES]
     38     grad = np.zeros_like(free)
---> 39     dJdq1 = np.pi*LATERAL_DISPLACEMENT*(LATERAL_DISPLACEMENT*(1 - cos(np.pi*q1/LONGITUDINAL_DISPLACEMENT))/2 - q2)*sin(pi*q1/LONGITUDINAL_DISPLACEMENT)/LONGITUDINAL_DISPLACEMENT
     40     dJdq2 = LATERAL_DISPLACEMENT*(cos(pi*q1/LONGITUDINAL_DISPLACEMENT) - 1) + 2*q2
     41     grad[:NUM_NODES] = dJdq1

```

**Comment by moorepants** [comment]
The constraints and jacobian seem to evaluate:

```
In [2]: problem.con(initial_guess)
Out[2]: array([0., 0., 0., ..., 0., 0., 0.])
In [4]: problem.jacobian(initial_guess)
Out[4]: array([ 0.5,  0. ,  0. , ...,  1. , -1. ,  1. ])
```

**Comment by moorepants** [comment]
Also not bad speed for 10k lines of c code:

```
In [5]: %timeit problem.jacobian(initial_guess)
2.03 ms ± 116 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by brocksam** [comment]
> The error seems to be in the objective:
> 
> ```
> EXIT: Invalid number in NLP function or derivative detected.
> ---------------------------------------------------------------------------
> NameError                                 Traceback (most recent call last)
> File ~/Manuscripts/muscle-driven-bicycle-paper/src/opty_ocp.py:152
>     149 initial_guess = np.zeros(problem.num_free)
>     151 # Find the optimal solution.
> --> 152 sol, info = problem.solve(initial_guess)
>     154 q1_sol = sol[:NUM_NODES]
>     155 q2_sol = sol[NUM_NODES:2*NUM_NODES]
> 
> File ~/miniconda/envs/muscle-bike-paper/lib/python3.11/site-packages/cyipopt/cython/ipopt_wrapper.pyx:642, in ipopt_wrapper.Problem.solve()
> 
> File ~/miniconda/envs/muscle-bike-paper/lib/python3.11/site-packages/cyipopt/cython/ipopt_wrapper.pyx:700, in ipopt_wrapper.gradient_cb()
> 
> File ~/src/opty/opty/direct_collocation.py:226, in Problem.gradient(self, free)
>     203 """Returns the value of the gradient of the objective function given a
>     204 solution to the problem.
>     205 
>    (...)
>     223 
>     224 """
>     225 # This should return a column vector.
> --> 226 return self.obj_grad(free)
> 
> File ~/Manuscripts/muscle-driven-bicycle-paper/src/opty_ocp.py:39, in obj_grad(free)
>      37 q2 = free[NUM_NODES:2*NUM_NODES]
>      38 grad = np.zeros_like(free)
> ---> 39 dJdq1 = np.pi*LATERAL_DISPLACEMENT*(LATERAL_DISPLACEMENT*(1 - cos(np.pi*q1/LONGITUDINAL_DISPLACEMENT))/2 - q2)*sin(pi*q1/LONGITUDINAL_DISPLACEMENT)/LONGITUDINAL_DISPLACEMENT
>      40 dJdq2 = LATERAL_DISPLACEMENT*(cos(pi*q1/LONGITUDINAL_DISPLACEMENT) - 1) + 2*q2
>      41 grad[:NUM_NODES] = dJdq1
> 
> NameError: name 'cos' is not defined
> > /home/moorepants/Manuscripts/muscle-driven-bicycle-paper/src/opty_ocp.py(39)obj_grad()
>      37     q2 = free[NUM_NODES:2*NUM_NODES]
>      38     grad = np.zeros_like(free)
> ---> 39     dJdq1 = np.pi*LATERAL_DISPLACEMENT*(LATERAL_DISPLACEMENT*(1 - cos(np.pi*q1/LONGITUDINAL_DISPLACEMENT))/2 - q2)*sin(pi*q1/LONGITUDINAL_DISPLACEMENT)/LONGITUDINAL_DISPLACEMENT
>      40     dJdq2 = LATERAL_DISPLACEMENT*(cos(pi*q1/LONGITUDINAL_DISPLACEMENT) - 1) + 2*q2
>      41     grad[:NUM_NODES] = dJdq1
> ```

Have you pulled the latest commits? I should have fixed this with https://github.com/brocksam/muscle-driven-bicycle-paper/pull/5/commits/7b95822163297b870760a23dcd8f2e8cfdde0b36 last night.

**Comment by brocksam** [comment]
Both `problem.con(initial_guess)` and `problem.jacobian(initial_guess)` produce NaN entries if an initial guess of all zeros is used. OCP runs with a non-zero initial guess. Good news. Ipopt quick starts doing restoration steps and then exceed max iterations, but that's to be expected with such a terrible initial guess. I'll work on producing a better one and hopefully we get somewhere.

**Comment by moorepants** [comment]
Can you push the latest up for this? There were some edits we made together that aren't yet there.

**Comment by brocksam** [comment]
> Can you push the latest up for this? There were some edits we made together that aren't yet there.

Done.

**Pull request merged**